### PR TITLE
Moving request_parser logic into Request

### PR DESCRIPTION
**Pull request by crystalhsieh**
Using the class's common API for Request, I moved the parser logic.

### CODE REVIEW

File: include/request.h [hunk 1]
```diff
@@ -3,22 +3,86 @@
...
+    // For testing. Must be removed to comply with API
+    enum result_type { good, bad, indeterminate };
+
+  private:
+
+
+    // Handle the next character of input.
+    result_type consume(char input);
+
+    std::string m_raw_request;
```
Comment (ryanvoong) [inline include/request.h]: change to raw_request_ (trailing underscore). same for the rest of the member variables

File: include/request.h [hunk 2]
```diff
@@ -3,22 +3,86 @@
...
+    std::string uri() const;
+    std::string version() const;
+
+    using Headers = std::vector<std::pair<std::string, std::string>>;
+    Headers headers() const;
+
+    std::string body() const;
+
+    // For testing. Must be removed to comply with API
+    enum result_type { good, bad, indeterminate };
```
Comment (ryanvoong) [inline include/request.h]: can this be private?

File: include/request.h [hunk 3]
```diff
@@ -3,22 +3,86 @@
...
+
+  private:
+
+
```
Comment (ryanvoong) [inline include/request.h]: too many newlines

File: src/request.cc [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+std::string Request::version() const {
+  return m_version;
+}
+
+Request::Headers Request::headers() const {
+  return m_headers;
+}
+
+// TODO: Mm... this shouldn't work.
+// What do they mean by body? Body of the HTTP request .. ?
```
Comment (ryanvoong) [inline src/request.cc]: the HTTP request will have a body if it's a POST request (which our server currently doesn't support). but basically just dump everything after the headers into the body_

File: include/request.h [hunk 5]
```diff
@@ -3,22 +3,86 @@
...
+    Headers m_headers;
+
+    // optional?
+    std::string m_handler_path;
+    std::string m_file_path;
+    int m_http_version_major;
+    int m_http_version_minor;
+
+    // Check if a byte is an HTTP character.
+    static bool is_char(int c);
```
Comment (ryanvoong) [inline include/request.h]: why is this static?

File: src/request.cc [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+std::string Request::method() const {
+  return m_method;
+}
+
+std::string Request::uri() const {
+  return m_uri;
+}
+
+// TODO: Not quite sure what they mean by "version"
+// Right now I just gather HTTP version major and minor
```
Comment (ryanvoong) [inline src/request.cc]: HTTP version, i.e. "1.1"
so it would be `http_version_major + "." + http_version_minor`

File: src/request.cc [hunk 7]
```diff
@@ -0,0 +1,275 @@
...
+
+Request::Headers Request::headers() const {
+  return m_headers;
+}
+
+// TODO: Mm... this shouldn't work.
+// What do they mean by body? Body of the HTTP request .. ? 
+std::string Request::body() const {
+  return "";
+}
```
Comment (ryanvoong) [inline src/request.cc]: newline

File: src/request.cc [hunk 8]
```diff
@@ -0,0 +1,275 @@
...
+  for (char c : raw_request) {
+    parsed_request->consume(c);
+  }
+  return parsed_request;
+}
+
+Request::result_type Request::consume(char input) {
+  m_raw_request.push_back(input);
+
+  switch (m_state) {
```
Comment (ryanvoong) [inline src/request.cc]: I think each case is supposed to be enclosed in curly braces
https://google.github.io/styleguide/cppguide.html#Loops_and_Switch_Statements

**Review by ryanvoong — CHANGES_REQUESTED** [review]

File: include/request.h (hunk 3 above)
Comment (crystalhsieh) [reply include/request.h]: done

File: include/request.h (hunk 1 above)
Comment (crystalhsieh) [reply include/request.h]: done

File: include/request.h (hunk 5 above)
Comment (crystalhsieh) [reply include/request.h]: this doesn't require an instance to be called ? It just checks if it's a character ?

File: src/request.cc (hunk 8 above)
Comment (crystalhsieh) [reply src/request.cc]: was a pain, but i did it lol

File: src/request.cc (hunk 6 above)
Comment (crystalhsieh) [reply src/request.cc]: done

Comment (crystalhsieh) [reply src/request.cc]: (implementation is in case http_version_minor_start and http_version_minor)

File: src/request.cc (hunk 4 above)
Comment (crystalhsieh) [reply src/request.cc]: quick fix where it's categorized as a "header" - talked about it offline about potential improvements

File: src/request.cc (hunk 7 above)
Comment (crystalhsieh) [reply src/request.cc]: done

File: src/request.cc [hunk 9]
```diff
@@ -0,0 +1,306 @@
...
+#include "request.h"
+
+// Parse some data. The enum return value is good when a complete request has
+// been parsed, bad if the data is invalid, indeterminate when more data is
+// required. The InputIterator return value indicates how much of the input
+// has been consumed.
+std::unique_ptr<Request> Request::Parse(const std::string& raw_request) {
+  std::unique_ptr<Request> parsed_request(new Request());
+  for (char c : raw_request) {
+    parsed_request->parsed_status_ = parsed_request->consume(c);
```
Comment (crystalhsieh) [inline src/request.cc:13]: This is kind of ugly but the other two options are:

1. make a wrapper method that sets parsed_status_ = parsed_request->consume(c);
    (so the wrapper method would call consume and set status)

2. make consume() return void, with it setting parsed_status_ and also setting return; (cuz it's a switch statement)

**Review by ryanvoong — APPROVED** [review]

**Pull request merged**